cns/deviceplugin/pluginmanager_test.go

```diff
@@ -0,0 +1,283 @@
+package deviceplugin_test
+
+import (
+	"context"
+	"net"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns/deviceplugin"
+	"github.com/Azure/azure-container-networking/crd/multitenancy/api/v1alpha1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+func TestPluginManagerStartStop(t *testing.T) {
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		t.Fatalf("error getting logger: %v", err)
+	}
+
+	// start up the fake kubelet
+	fakeKubeletSocketDir := os.TempDir()
+	kubeletSocket := path.Join(fakeKubeletSocketDir, "kubelet.sock")
+	kubeletErrChan := make(chan error)
+	vnetPluginRegisterChan := make(chan string)
+	ibPluginRegisterChan := make(chan string)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		kubeletErr := runFakeKubelet(ctx, kubeletSocket, vnetPluginRegisterChan, ibPluginRegisterChan, fakeKubeletSocketDir)
+		kubeletErrChan <- kubeletErr
+	}()
+
+	// run the plugin manager
+	expectedVnetNICs := 2
+	expectedIBNICs := 3
+	manager := deviceplugin.NewPluginManager(logger, expectedVnetNICs, expectedIBNICs,
+		deviceplugin.PluginManagerSocketPrefix(fakeKubeletSocketDir),
+		deviceplugin.PluginManagerKubeletSocket(kubeletSocket),
+		deviceplugin.PluginDeviceCheckInterval(time.Second))
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- manager.Run(ctx)
+	}()
+
+	// wait till the two plugins register themselves with fake kubelet
+	vnetPluginEndpoint := <-vnetPluginRegisterChan
+	ibPluginEndpoint := <-ibPluginRegisterChan
+
+	// assert the plugin reports the expected vnet nic count
+	gotVnetNICCount := getDeviceCount(t, vnetPluginEndpoint)
+	if gotVnetNICCount != expectedVnetNICs {
+		t.Fatalf("expected %d vnet nics but got %d", expectedVnetNICs, gotVnetNICCount)
+	}
+	gotIBNICCount := getDeviceCount(t, ibPluginEndpoint)
+	if gotIBNICCount != expectedIBNICs {
+		t.Fatalf("expected %d ib nics but got %d", expectedIBNICs, gotIBNICCount)
+	}
+
+	// update the device counts and assert they match expected after some time
+	expectedVnetNICs = 5
+	expectedIBNICs = 6
+	if err = manager.TrackDevices(v1alpha1.DeviceTypeVnetNIC, expectedVnetNICs); err != nil {
+		t.Fatalf("error tracking VNET devices: %v", err)
+	}
+
+	if err = manager.TrackDevices(v1alpha1.DeviceTypeInfiniBandNIC, expectedIBNICs); err != nil {
+		t.Fatalf("error tracking IB devices: %v", err)
+	}
+
+	<-time.After(3 * time.Second)
```

Comment:
Any way to remove this (ideally) ? Makes me feel like a source of future flakiness and I'm allergic to slow tests :)

Reply:
We want to wait enough time before the nic update is reflected. Making it smaller could actually make the test flaky.

Reply:
You can probably reduce the device update poller (see above `deviceplugin.PluginDeviceCheckInterval(time.Second)`) to something like 500ms and then wait on 1 second instead.

Reply:
Or, you could call `getDeviceCount` in a `retrier` until you get the expected result (or exceed some timeout threshold)

Reply:
This is usually because there are temporal dependencies that aren't adequately addressed. Usually this can be fixed by introducing a channel or some function to be used as a latch. Time is an illusion, and what might be an adequate sleep today might not be tomorrow. It is worth fighting to remove these.

Reply:
I think in this case, the "black box" that we're testing is that a client can call TrackDevices, and _eventually_ `getDeviceCount` will reflect that new value. But we don't really know or make any guarantees on what eventually means. So I think retrying until some reasonable threshold is probably the most accurate way of testing the interface, I think injecting a function as a latch is kind of coupling the test and the implementation, in this case.

Reply:
updated to 6 retry attempts with 500ms delay in between.

Reply:
Yeah, I think that's more reasonable. At least there's a chance it could resolve sooner. Also, @thatmattlong , I must have written my comment concurrently, since I didn't see your input here when I posted.